### PR TITLE
Deduplicated metatables

### PR DESCRIPTION
**Pull request by FabricatorZayac**
Made it so metatables for shaders and matrices are created during initialization and referenced each time a matrix/shader object are created instead of creating a new metatable for each new object

### CODE REVIEW

File: src/model.c [hunk 1]
```diff
@@ -66,6 +64,8 @@ int ir_model_new(ir_model *model) {
 
 	ir_vector_init_lua(model->state);
 
+  ir_shader_init_lua(model->state);
```
Comment (cbpudding) [inline src/model.c]: Not sure how this line changed. Seems like a minor thing but it's still worth mentioning for readability's sake.

File: src/shader.c [hunk 2]
```diff
@@ -144,20 +169,9 @@ int ir_shader_new_lua(lua_State *L) {
...
-	lua_pushstring(L, "shader");
-	lua_setfield(L, -2, "__type");
-
-	lua_pushcfunction(L, ir_shader_use_lua);
-	lua_setfield(L, -2, "use");
-
-	lua_setfield(L, -2, "__index");
+    lua_getglobal(L, "ir");
+    lua_getfield(L, -1,  "shader_meta");
+    lua_replace(L, -2);
```
Comment (cbpudding) [inline src/shader.c]: I think we should standardize on ir.thing.meta to keep the top level table clean.

**Review by cbpudding — CHANGES_REQUESTED** [review]

File: src/shader.c (hunk 2 above)
Comment (FabricatorZayac) [reply src/shader.c]: Sure. But `ir.shader` is a function, unlike `ir.matrix`, which is a table, so I don't know what to do there

Comment (FabricatorZayac) [reply src/shader.c]: We could have it be `ir.meta.thing` instead

File: src/model.c (hunk 1 above)
Comment (FabricatorZayac) [reply src/model.c]: I moved out the binding of the `ir_shader_new_lua` function into a separate `ir_shader_init_lua` function together with the initialization of the shader metatable. That line wasn't there

Comment (cbpudding) [reply src/model.c]: Ah. That'll do it. It should be fixed on the next `clang-format`.

File: src/shader.c (hunk 2 above)
Comment (cbpudding) [reply src/shader.c]: Shooting myself in the foot again I see... Maybe it would be better if we moved both under `ir.internal.meta.*`?

Comment (FabricatorZayac) [reply src/shader.c]: Sounds good to me. On it

File: src/model.c [hunk 3]
```diff
@@ -95,155 +96,195 @@ int ir_model_time_lua(lua_State *L) {
...
 
-	lua_pushinteger(model->state, ALLEGRO_EVENT_TOUCH_BEGIN);
-	lua_setfield(model->state, -2, "EVENT_TOUCH_BEGIN");
+    lua_setfield(L, -2, "shader");
+    // /shader_meta
 
-	lua_pushinteger(model->state, ALLEGRO_EVENT_TOUCH_END);
-	lua_setfield(model->state, -2, "EVENT_TOUCH_END");
+    // mat_meta
+	lua_createtable(L, 0, 3);
```
Comment (cbpudding) [inline src/model.c:270]: Why not keep `ir` on the stack and have it be referenced twice? Probably not a big deal since this is one time initialization.

**Review by cbpudding — APPROVED** [review]

**Pull request merged**